### PR TITLE
kubeadm: graduate the control-plane phase

### PR DESCRIPTION
**Pull request by neolit123**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

kubeadm: move the controlplane phase to phase runner.

Follows the example of recently the implemented preflight phase.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Refs https://github.com/kubernetes/kubeadm/issues/1163

**Special notes for your reviewer**:

outdated info:
<details>
@fabriziopandini the sub-phases of controlplane need to add flags.
PTAL at how i did it, but i don't like it that much.

- first i had to pass the internal config to the function NewControlPlanePhase() in NewCmdInit()
- i had to add Flags() to the Phase structs
- these flags are combined with the command flags.
</details>

side note: removed controlplane_test.go as it feels redundant now.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @fabriziopandini 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/kind feature
/area kubeadm

depends on:
https://github.com/kubernetes/kubernetes/pull/70331
https://github.com/kubernetes/kubernetes/pull/70098

### CODE REVIEW

**Comment by neolit123** [comment]
/hold
(bots not working now).

but this PR depends on a feature PR about the kubeadm  kubeconfig phase.

File: cmd/kubeadm/app/cmd/init.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/init.go]: InitData are run time data; it confuses me to get those data passed when cmd are created (and I'm not 100% sure this will behave the same)

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: Those flags should be dropped

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: If this phase is only a container for sub phases you could omit Run

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/controlplane.go]: By dropping those flags, you don't need cfg anymore

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: If you check at the PR I Iinked, I'm proposing a slightly different approach for doing this.
WDYT?

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: If you check to PR I linked, I'm proposing an approach that will allow you to get rid of all the Create* wrappers in favor of createStaticPodFiles

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@neolit123 thanks for sharing this WIP.
I don't think we need flags, because flags of phases should be (<)= of flags of the command they belong to (this was discussed in the KEP and if I remember well also in the PR for the phase runner.

In this case, IMO we should drop extra args flags in favor of the config file.

With regards to the implementation, I think there is room for a little bit more of refactoring/clean-up; see e.g. https://github.com/kubernetes/kubernetes/pull/70098 last two commits & comments in line

**Comment by neolit123** [comment]
@fabriziopandini 
i think i've addressed all comments.
removing the flags for sub-phases made things much cleaner.

this PR also now moves the audit policy in the API server sub-phase. PTAL.

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: this code is bit difficult to tackle with phases as is.

it first does a:
`realCertsDir := i.cfg.CertificatesDir`
on line 446 it does a `i.cfg.CertificatesDir = certsDirToWriteTo`
and here it goes back to`i.cfg.CertificatesDir = realCertsDir`

i think my change is breaking here; i need to verify that.

comments appreciated.

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@neolit123. Well done!
You can find inline the answer to your question + some nits
IMO you can now remove WIP now so we can get more eyes on this.

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/init.go]: ok, understood.

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: reading the code here it seems to me that in the case of etcd and control plane the dry run folder is not used and instead the real (value from config) value is used.

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: seems to me that the original code in runInit() used the same folder for dry-run and without dry-run in the case of etcd and api server.

here by using ManifestDir() we now respect the dry-run flag.

**Comment by neolit123** [comment]
@fabriziopandini manged to update the PR today.
PTAL.

**Comment by fabriziopandini** [comment]
/approve
@neolit123 IMO now we are ready (but please keep hold until other phases gets merged)
The problem in dry run was fixed in kubelet-start phase, so if you rebase your PR you will get integration test passing. It seems you need also tu run upgrade-generared -docs.sh to pass kubernetes verify

**Comment by neolit123** [comment]
> @neolit123 IMO now we are ready (but please keep hold until other phases gets merged)
The problem in dry run was fixed in kubelet-start phase, so if you rebase your PR you will get integration test passing.

sounds good.

> It seems you need also tu run upgrade-generared -docs.sh to pass kubernetes verify

sadly my PR that nukes the placeholder docs is blocked by a "riddle" from a maintainer, with not further clarifications.

File: cmd/kubeadm/app/cmd/phases/kubeconfig_test.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubeconfig_test.go]: this seems to be a constant that we are passing around we probably should just clean it up later.
_test.go files depend on it.

**Comment by neolit123** [comment]
@fabriziopandini updated the PR, fixed the integration issue too.

output with -v=0 is:
```
[control-plane] creating static Pod manifest for kube-apiserver
[control-plane] creating static Pod manifest for kube-controller-manager
[control-plane] creating static Pod manifest for kube-scheduler
````
-v=1 shows the path where files are written.

will re-visit again tomorrow if needed.

File: cmd/kubeadm/app/cmd/init.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/init.go]: @neolit123 If I'm not wrong now it possible to nuke set/reset of `certificateDir` from `runInit`

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: in certs and kubeconfig I have added a run method to the top level phase that prints the folder where stuff are going to be saved. What do you think about doing the same here?

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplane.go]: %s/%q for consistency with other phases

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/controlplane.go]: I don't have history here, but what you are proposing seems ok to me.

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 ready to go!
Some minor that can be fixed now or later when we will do a final review on UX

/approve
/lgtm

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/init.go]: i'm removing it in the etcd phase PR but we can do it here.

File: cmd/kubeadm/app/cmd/phases/controlplane.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/controlplane.go]: SGTM

**Comment by neolit123** [comment]
@fabriziopandini addressed latest feedback.

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 Thanks
/approve
/lgtm

**Comment by neolit123** [comment]
@fabriziopandini 
will rebase the etcd phase later today.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**
